nexus/saga-recovery/src/lib.rs

```diff
@@ -0,0 +1,659 @@
+//! we're trying to do here.  The rest of this comment assumes you've read all
+//! that.
+//!
+//! ## Saga recovery passes
+//!
+//! For the reasons mentioned in that comment, saga recovery is done in
+//! **passes**.  The state that's kept between passes (when the task is "at
+//! rest") is called the [`RestState`].
+//!
+//! Each saga recovery pass looks like this:
+//!
+//! 1. Start with initial [`RestState`] and [`Report`].
+//! 2. List in-progress sagas from the database.
+//! 3. Collect list of sagas that have been started by the rest of Nexus.
+//! 4. Use [`Plan::new()`] to construct a plan.  The plan examines all the sagas
+//!    reported by the database as well as all the sagas we knew about before
+//!    and determines for each one exactly what's needed.  Each saga falls into
+//!    one of a few buckets:
+//!    * It's clearly not running, but should be, so it needs to be recovered.
+//!    * It's clearly running, so it does not need to be recovered.
+//!    * It has finished running and we can stop keeping track of it.
+//!    * It _may_ be running but we cannot tell because of the intrinsic race
+//!      between steps 2 and 3 above.  We'll keep track of these and resolve the
+//!      ambiguity on the next pass.
+//! 5. Carry out recovery for whatever sagas need to be recovered.  Use
+//!    [`ExecutionBuilder::new()`] to construct a description of what happened.
+//! 6. Update the [`RestState`] and [`Report`] to reflect what happened in this
+//!    pass.
+//!
+//! This process can be repeated forever, as often as wanted, but should not be
+//! run concurrently.
+//!
+//! ## Saga recovery task vs. this crate
+//!
+//! This process is driven by the caller (the saga recovery background task),
+//! with helpers provided by this crate:
+//!
+//! ```text
+//!          Saga recovery task        |      nexus-saga-recovery crate
+//!     ------------------------------------------------------------------------
+//!                                    |
+//!      1. initial `RestState` and   ---> provides `RestState`, `Report`
+//!         `Report`                   |
+//!                                    |
+//!      2. list in-progress sagas     |
+//!                                    |
+//!      3. collect list of sagas  ------> use RestState::update_started_sagas()
+//!         started by Nexus           |
+//!                                    |
+//!      4. make a plan  ----------------> use Plan::new()
+//!                                    |   This is where all the decisions
+//!                                    |   about saga recovery get made.
+//!                                    |
+//!      5. follow the plan -------------> use Plan::sagas_needing_recovery()
+//!                                    |
+//!         fetch details from db      |
+//!         load sagas into Steno      |
+//!                                    |   use ExecutionBuilder::new() to report
```

Comment:
_very_ nitpicky, sorry: you've put backticks around _some_ of the function/type names, here but not all of 'em:
```suggestion
//!      3. collect list of sagas  ------> use `RestState::update_started_sagas()`
//!         started by Nexus           |
//!                                    |
//!      4. make a plan  ----------------> use `Plan::new()`
//!                                    |   This is where all the decisions
//!                                    |   about saga recovery get made.
//!                                    |
//!      5. follow the plan -------------> use `Plan::sagas_needing_recovery()`
//!                                    |
//!         fetch details from db      |
//!         load sagas into Steno      |
//!                                    |   use `ExecutionBuilder::new()` to report
```